### DISCUSSION TITLE
Using nanopb in configurable library outside of application folder

### DISCUSSION THREAD

**Discussion by asteriskSF** [zephyrproject-rtos/zephyr#77494] [Q&A]
I am working on a project with multiple build variations off a common code base.  All of the build variations use protobuf with nanopb but some of the protobuf files do not apply to some of the build variations.   As an example there are multiple boards supported, each with multiple applications such as a manufacturing test application, and various customer-focused applications for different markets.  

The general approach has been to place all these configurable library/modules into a folder where the CmakeLists.txt is conditionally included based on a Kconfig boolean flag  such as CONFIG_LIB_OPTIONAL_PROTOBUF_COMPONENT1  Each of these conditional directories builds its files in a zephyr library, so the cmake usually has:

zephyr_include_directories(include)
zephyr_library()
zephyr_library_sources(library_source_file1.c)

The nanopb integration seems to be focused solely on protobuf files found directly in the application folder.  This doesn't work in my application because the .proto files are share across multiple applications.  In addition, they are actually placed into a submodule so the proto files can be shared with the mobile app developers.

I'm looking for advise on how to craft the configurable library's cmake file to:
* run protoc to convert the .proto into   *.c and *.h
* compile *.c into a the zephyr_library()
* add the current directory into the include list

Also what should I have in the application cmake to trigger building the common nanopb files, although there are not any application specific protobuf files.

thanks for any help!

**Comment by asteriskSF** [comment]
@pdgendt Maybe you can give me some help, I see you have some experience with the nanopb targets!

**Reply by pdgendt** [reply]
Take a look at [tests/modules/nanopb/lib/CMakeLists.txt](https://github.com/zephyrproject-rtos/zephyr/blob/143b14bb4fb779437fdeb8369e42fbb81db4a5fd/tests/modules/nanopb/lib/CMakeLists.txt), it demonstrates using the generated headers in a library.

**Comment by asteriskSF** [comment]
@pdgendt Thanks for your response..

From what I see, that example doesn't match what I'm trying to create.

I have multiple application folders.  I also have multiple library folders with .proto files  For maintainability, the code is structured with lib, boards, drivers folders outside of the application folder.  This is based on the zephyr example_application: https://github.com/zephyrproject-rtos/example-application

I'm trying to create a structure like this;

lib/CMakeLIsts.txt
lib/Kconfig
lib/schemas/CMakeLIsts.txt
lib/schemas/Kconfig
lib/schemas/schema1/CMakeLists.txt
lib/schemas/schema1/Kconfig
lib/schemas/schema1/schema1.proto
lib/schemas/schema1/schema1.pb.c    //generated files
lib/schemas/schema1/schema1.pb.h    //generated files
lib/schemas/schema2/CMakeLists.txt
lib/schemas/schema2/Kconfig
lib/schemas/schema2/schema2.proto
lib/schemas/schema2/schema2.pb.c      //generated files
lib/schemas/schema2/schema2.pb.h      //generated files

The application directories are parallel to the lib/schema folders:

applications/mfg_app/CMakeLists.txt
applications/mfg_app/prj.conf
applications/mfg_app/src
applications/product_app/CMakeLists.txt
application/product_app/CmakeLists.txt

The application/prj.conf would only have a single additional line:
CONFIG_SCHEMA1=y 
which would cause schema1 proto/c/h files to be automatically added into the build.

How do I add proto targets to generate the .c and .h files in the lib/schemas/schema1/CMakeLists.txt?

Perhaps we could make a PR to add a nanopb library to the example_application repo?

a*

**Reply by pdgendt** [reply]
I think you want to add the following in `lib/schemas/schema1/CMakeLists.txt` (and similar for `schema2`)

```cmake
zephyr_library() # this creates a schema1 library target

# Make sure the c/h files are generated
zephyr_nanopb_sources(schema1
  schema1.proto
)

# Add the path of the generated header file to the global includes
zephyr_include_directories($CMAKE_CURRENT_BINARY_DIR)
```

In any of the application CMake files make sure to add a dependency on nanopb generated headers (so you can include them), e.g.
```cmake
cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(mfg_app)

list(APPEND CMAKE_MODULE_PATH ${ZEPHYR_BASE}/modules/nanopb)
include(nanopb)

FILE(GLOB app_sources src/*.c)
target_sources(app PRIVATE ${app_sources})

if(CONFIG_SCHEMA1)
  # Make sure the headers are generated before building app
  add_dependencies(app nanopb_generated_headers)

  add_subdirectory(lib/schemas/schema1)
  target_link_libraries(app PRIVATE schema1)
endif()
```

**Comment by asteriskSF** [comment]
@pdgendt This is helping, I made some progress but not quite there yet.

I did have to make a change, the library name from zephyr_library is complicated but its available from a variable $ZEPHYR_CURRENT_LIBRARY.

I put this is the lib/schema1/CMakeLists.txt:


```zephyr_library()

zephyr_include_directories($CMAKE_CURRENT_BINARY_DIR)
zephyr_nanopb_sources(${ZEPHYR_CURRENT_LIBRARY} schema1.proto)
```

I also put this in the app/CMakeLists.txt
```
project(app LANGUAGES C)

list(APPEND CMAKE_MODULE_PATH ${ZEPHYR_BASE}/modules/nanopb)
include(nanopb)

target_sources(app PRIVATE src/main.c)

add_dependencies(app nanopb_generated_headers)
```

I didn't  put the CONFIG_SCHEMA1  in the app directory since that is included from lib\CMakeLists.txt based on a Kconfig from lib\schema1\Kconfig

However I still have a problem.

I have an additional library we'll call it lib1, which provides functions using schema1.  Its header file lib1.h includes schema1.pb.h..

When I compile main.c which include lib1.h and schema1.h

In file included from XXX/app/src/main.c:16:
XXX/lib/lib1/include/lib1.h:14:10: fatal error: schema1.pb.h: No such file or directory
   14 | #include "schema1.pb.h"
      |          ^~~~~~~~~~~~~~~~~~
compilation terminated.

It seems like main.c is compiled before hud_control.pb.h is generated or possibly the include directory is not setup correctly.

I did find the schema1.pb.h is generated in the build directory
find ./ -name "schema1*" -print
./build/modules/app.git/lib/schema1
./build/modules/app.git/lib/schema1/schema1.pb.c
./build/modules/app.git/lib/schema1/schema1.pb.h

Any ideas?

**Reply by pdgendt** [reply]
Any CMake target that includes a generated nanopb header needs to add
```cmake
add_dependencies(<target> nanopb_generated_headers)
```
That should ensure that the headers are generated before compiling any source in the target. Also note that `zephyr_include_directories` works for `zephyr` targets, but not for "vanilla" CMake targets.

You could try to add the following in `schema1/CMakeLists.txt` instead
```cmake
include_directories($CMAKE_CURRENT_BINARY_DIR)
```

**Comment by asteriskSF** [comment]
I think I have it working now.

In the lib/schema1/CMakeLists.txt I have:
```
list(APPEND CMAKE_MODULE_PATH ${ZEPHYR_BASE}/modules/nanopb)
include(nanopb)

zephyr_library()
zephyr_include_directories(${CMAKE_CURRENT_BINARY_DIR})
zephyr_nanopb_sources(${ZEPHYR_CURRENT_LIBRARY} schema1.proto)
```
In my previous code I had forgotten the braces around CMAKE_CURRENT_BINARY_DIR

**Reply by pdgendt** [reply]
Hey @asteriskSF, if you think this is fixed, can you close the discussion?

**Comment by asteriskSF** [comment]
I have it working, thanks for the help!
